File: platypus-js-web-client/src/main/java/com/eas/grid/Section.java
package com.eas.grid;

import com.eas.bound.JsArrayList;
import com.eas.core.XElement;
import com.eas.grid.columns.Column;
import com.eas.grid.columns.header.HeaderNode;
import java.util.List;

import com.google.gwt.core.client.JavaScriptObject;
import com.google.gwt.dom.client.BrowserEvents;
import com.google.gwt.dom.client.Document;
import com.google.gwt.dom.client.Element;
import com.google.gwt.dom.client.NativeEvent;
import com.google.gwt.dom.client.NodeList;
import com.google.gwt.dom.client.TableCellElement;
import com.google.gwt.dom.client.TableColElement;
import com.google.gwt.dom.client.TableElement;
import com.google.gwt.dom.client.TableRowElement;
import com.google.gwt.dom.client.TableSectionElement;
import com.google.gwt.event.dom.client.KeyCodes;
import java.util.ArrayList;

/**
 *
 * @author mg
 */
public class Section {

    public static final String JS_ROW_NAME = "js-row";

    private TableElement table = Document.get().createTableElement();
    private TableColElement colgroup = Document.get().createColGroupElement();
    private List<HeaderNode> header = new ArrayList<>();
    private Element keyboardSelectedElement;
    protected String dynamicCellClassName;
    protected String dynamicOddRowsClassName;
    protected String dynamicEvenRowsClassName;
    protected String dynamicHeaderRowClassName;
    private boolean draggableRows;
    private int keyboardSelectedRow = -1;
    private int keyboardSelectedColumn = -1;
    private int rowsPerPage = 30; // Only for PageDown or PageUp keys handling
    private JsArrayList data; // Already sorted
    /**
     * Already sorted data array indices;
     */
    private int startRow; // Inclusive
    private int endRow; // Exclusive
    private List<Column> columns = new ArrayList<>();

    public Section(String aDynamicCellClassName, String aDynamicOddRowsClassName, String aDynamicEvenRowsClassName, String aDynamicHeaderRowClassName) {
        super();
        this.dynamicCellClassName = aDynamicCellClassName;
        this.dynamicOddRowsClassName = aDynamicOddRowsClassName;
        this.dynamicEvenRowsClassName = aDynamicEvenRowsClassName;
        this.dynamicHeaderRowClassName = aDynamicHeaderRowClassName;
        TableSectionElement thead = table.getTHead();
        thead.appendChild(colgroup);

        table.getStyle().setProperty("borderCollapse", "collapse");
        table.<XElement>cast().addEventListener(BrowserEvents.KEYDOWN, new XElement.NativeHandler() {
            @Override
            public void on(NativeEvent event) {
                int oldRow = Section.this.keyboardSelectedRow;
                int oldColumn = Section.this.keyboardSelectedColumn;
                int keyCode = event.getKeyCode();
                if (keyCode == KeyCodes.KEY_LEFT) {
                } else if (keyCode == KeyCodes.KEY_RIGHT) {
                } else if (keyCode == KeyCodes.KEY_UP) {
                } else if (keyCode == KeyCodes.KEY_DOWN) {
                } else if (keyCode == KeyCodes.KEY_PAGEUP) {
                } else if (keyCode == KeyCodes.KEY_PAGEDOWN) {
                } else if (keyCode == KeyCodes.KEY_HOME) {
                } else if (keyCode == KeyCodes.KEY_END) {
                }
            }
        });
    }

    public TableElement getElement() {
        return table;
    }

    public boolean isDraggableRows() {
        return draggableRows;
    }

    public void setDraggableRows(boolean aValue) {
        if (draggableRows != aValue) {
            draggableRows = aValue;
        }
    }

    public int getKeyboardSelectedColumn() {
        return keyboardSelectedColumn;
    }

    public void setKeyboardSelectedColumn(int keyboardSelectedColumn) {
        this.keyboardSelectedColumn = keyboardSelectedColumn;
    }

    public int getKeyboardSelectedRow() {
        return keyboardSelectedRow;
    }

    public void setKeyboardSelectedRow(int keyboardSelectedRow) {
        this.keyboardSelectedRow = keyboardSelectedRow;
    }

    public void addColumn(Column aColumn) {
        addColumn(aColumn, true);
    }

    public void addColumn(Column aColumn, boolean needRedraw) {
        columns.add(aColumn);
        colgroup.appendChild(aColumn.getElement());
        if (needRedraw) {
            redraw();
        }
    }

    public void addColumn(int index, Column aColumn) {
        addColumn(index, aColumn, true);
    }

    public void addColumn(int index, Column aColumn, boolean needRedraw) {
        if (index >= 0 && index <= columns.size()) { // It is all about insertBefore
            if (index < columns.size()) { // It is all about insertBefore
                columns.add(index, aColumn);
                Element col = (Element) colgroup.getChild(index);
                colgroup.insertBefore(aColumn.getElement(), col);
            } else {
                addColumn(aColumn);
            }
            if (needRedraw) {
                redraw();
            }
        }
    }

    public Column removeColumn(int index) {
        return removeColumn(index, true);
    }

    public Column removeColumn(int index, boolean needRedraw) {
        if (index >= 0 && index < columns.size()) {
            Column removed = columns.remove(index);
            removed.getElement().removeFromParent();
            removed.getColumnRule().removeFromParent();
            if (needRedraw) {
                redraw();
            }
            return removed;
        } else {
            return null;
        }
    }

    public int getColumnCount() {
        return columns.size();
    }

    public Column getColumn(int index) {
        return index >= 0 && index < columns.size() ? columns.get(index) : null;
    }

    public int getColumnIndex(Column column) {
        return columns.indexOf(column);
    }

    public int getRowsCount() {
        return table.getTBodies().getItem(0).getRows().getLength();
    }

    public TableCellElement getViewCell(int aRow, int aCol) {
        NodeList<TableRowElement> viewRows = table.getTBodies().getItem(0).getRows();
        if (aRow - startRow >= 0 && aRow - startRow < viewRows.getLength()) {
            TableRowElement viewRow = viewRows.getItem(aRow);
            NodeList<TableCellElement> cells = viewRow.getCells();
            if (aCol >= 0 && aCol < cells.getLength()) {
                return cells.getItem(aCol);
            }
        }
        return null;
    }

    public void focusCell(int aRow, int aCol) {
        setKeyboardSelectedColumn(aCol);
        setKeyboardSelectedRow(aRow);
        TableCellElement cell = getViewCell(aRow, aCol);
        cell.focus();
    }

    public Element getKeyboardSelectedElement() {
        return keyboardSelectedElement;
    }

    public void redraw() {
        redrawHeaders();
        redrawBody();
        redrawFooters();
    }

    public void redrawBody() {
        TableSectionElement tbody = table.getTBodies().getItem(0);
        tbody.removeFromParent();
        table.appendChild(Document.get().createTBodyElement());
        drawBody();
    }

    public void drawBody() {
        if (!columns.isEmpty()) {
            drawBodyPortion(startRow, endRow);
        }
    }

    private void drawBodyPortion(int start, int end) {
        TableSectionElement tbody = table.getTBodies().getItem(0);
        for (int i = start; i < end; i++) {
            JavaScriptObject dataRow = data.get(i);
            TableRowElement viewRow = Document.get().createTRElement();
            if ((i + 1) % 2 == 0) {
                viewRow.addClassName(dynamicEvenRowsClassName);
            } else {
                viewRow.addClassName(dynamicOddRowsClassName);
            }
            viewRow.addClassName("selected-row");
            viewRow.setPropertyJSO(JS_ROW_NAME, dataRow);
            if (i < startRow) {
                tbody.insertFirst(viewRow);
            } else {
                tbody.appendChild(viewRow);
            }
            for (int c = 0; c < columns.size(); c++) {
                Column column = columns.get(c);
                TableCellElement viewCell = Document.get().createTDElement();
                // TODO: Check alignment of the cell
                // TODO: Check image decoration of the cell and decoration styles
                viewCell.addClassName(dynamicCellClassName);
                viewRow.appendChild(viewCell);
                column.render(i, dataRow, viewCell);
            }
        }
    }

    public void setRange(int aStartRow, int aEndRow) {
        startRow = aStartRow;
        endRow = aEndRow;
        redrawBody();
    }

    public void shiftRange(int aStartRow, int aEndRow) {
        if (aStartRow >= endRow || aEndRow < startRow) {
            setRange(aStartRow, aEndRow);
        } else {
            TableSectionElement tbody = table.getTBodies().getItem(0);
            if (aStartRow < startRow) {
                drawBodyPortion(aStartRow, startRow);
            } else if (aStartRow > startRow) {
                for (int i = startRow; i < aStartRow; i++) {
                    tbody.getRows().getItem(0).removeFromParent();
                }
            }
            startRow = aStartRow;
            if (aEndRow < endRow) {
                for (int i = endRow - 1; i >= aEndRow; i--) {
                    tbody.getRows().getItem(tbody.getRows().getLength() - 1).removeFromParent();
                }
            } else if (aEndRow > endRow) {
                drawBodyPortion(endRow, aEndRow);
            }
            endRow = aEndRow;
        }
    }

    public void redrawHeaders() {
        TableSectionElement thead = table.getTHead();
        thead.removeFromParent();
        table.appendChild(Document.get().createTHeadElement());
        drawHeaders();
    }

    public void drawHeaders() {
        if (!columns.isEmpty()) {
        }
    }

    public void redrawFooters() {
        TableSectionElement tbody = table.getTFoot();
        tbody.removeFromParent();
        table.appendChild(Document.get().createTFootElement());
        drawFooters();
    }

    public void drawFooters() {
        if (!columns.isEmpty()) {
        }
    }

    public void redrawRow(int index) {
        TableSectionElement tbody = table.getTBodies().getItem(0);
        if (index - startRow >= 0 && index - startRow < tbody.getRows().getLength()) {
            TableRowElement viewRow = (TableRowElement) tbody.getRows().getItem(index - startRow);
            JavaScriptObject dataRow = data.get(index);
            for (int c = 0; c < columns.size(); c++) {
                Column column = columns.get(c);
                TableCellElement viewCell = (TableCellElement) viewRow.getCells().getItem(c);
                column.render(index, dataRow, viewCell);
            }
        } // if the row is not rendered then we have to do nothing
    }

    public void redrawColumn(int aIndex) {
        if (aIndex >= 0 && aIndex < columns.size()) {
            Column column = columns.get(aIndex);
            TableSectionElement tbody = table.getTBodies().getItem(0);
            for (int i = startRow; i < endRow; i++) {
                JavaScriptObject dataRow = data.get(i);
                TableRowElement viewRow = (TableRowElement) tbody.getRows().getItem(i - startRow);
                TableCellElement viewCell = (TableCellElement) viewRow.getCells().getItem(aIndex);
                while (viewCell.getFirstChildElement() != null) {
                    viewCell.getFirstChildElement().removeFromParent();
                }
                column.render(i, dataRow, viewCell);
            }
        }
    }

    public void setHeaderNodes(List<HeaderNode> aHeader) {
        setHeaderNodes(aHeader, true);
    }

    public void setHeaderNodes(List<HeaderNode> aHeader, boolean needReadraw) {
        header = aHeader;
        if (needReadraw) {
            redrawHeaders();
        }
    }

    public void clearColumnsAndHeader() {
        clearColumnsAndHeader(true);
    }

    public void clearColumnsAndHeader(boolean needRedraw) {
        for (int i = columns.size() - 1; i >= 0; i--) {
            Column removed = columns.remove(i);
            removed.getElement().removeFromParent();
            removed.getColumnRule().removeFromParent();
        }
        header = new ArrayList<>();
        if (needRedraw) {
            redraw();
        }
    }
}

File: platypus-js-web-client/src/app/forms/grid/section.js
define([
    '../../ui',
    '../../invoke',
    '../../logger',
    '../key-codes'
], function (
        Ui,
        Invoke,
        Logger,
        KeyCodes
        ) {

    var JS_ROW_NAME = "js-row";

    function Section(grid, dynamicCellsClassName,
            dynamicRowsClassName,
            dynamicHeaderCellsClassName,
            dynamicHeaderRowsClassName,
            dynamicOddRowsClassName,
            dynamicEvenRowsClassName) {
        var self = this;

        var table = document.createElement('table');
        table.className = 'p-grid-section';
        var colgroup = document.createElement('colgroup');
        var rowsHeight = 30;
        var headerNodes = [];
        var headerMaxDepth = 0;
        var keyboardSelectedElement = null;
        var draggableRows = false;
        var keyboardSelectedRow = -1;
        var keyboardSelectedColumn = -1;
        var rowsPerPage = 30; // Only for PageDown or PageUp keys handling
        var data = []; // Already sorted
        /**
         * Already sorted data array indices;
         */
        var startRow = 0; // Inclusive
        var endRow = 0; // Exclusive
        var columns = [];

        var thead;
        recreateHead();
        var tbody;
        recreateBody();
        var tfoot;
        recreateFoot();
        var bodyFiller = document.createElement('div');
        bodyFiller.className = 'p-grid-body-filler';

        Ui.on(table, Ui.Events.KEYDOWN, function (event) {
            var oldRow = self.keyboardSelectedRow;
            var oldColumn = self.keyboardSelectedColumn;
            var keyCode = event.keyCode;
            if (keyCode === KeyCodes.KEY_LEFT) {
            } else if (keyCode === KeyCodes.KEY_RIGHT) {
            } else if (keyCode === KeyCodes.KEY_UP) {
            } else if (keyCode === KeyCodes.KEY_DOWN) {
            } else if (keyCode === KeyCodes.KEY_PAGEUP) {
            } else if (keyCode === KeyCodes.KEY_PAGEDOWN) {
            } else if (keyCode === KeyCodes.KEY_HOME) {
            } else if (keyCode === KeyCodes.KEY_END) {
            }
        });


        Object.defineProperty(this, 'element', {
            get: function () {
                return table;
            }
        });
        Object.defineProperty(this, 'data', {
            get: function () {
                return data;
            },
            set: function (aValue) {
                if (data !== aValue) {
                    data = aValue;
                }
            }
        });
        Object.defineProperty(this, 'rowsHeight', {
            get: function () {
                return rowsHeight;
            },
            set: function (aValue) {
                if (rowsHeight !== aValue) {
                    rowsHeight = aValue;
                    redrawBody();
                }
            }
        });
        Object.defineProperty(this, 'draggableRows', {
            get: function () {
                return draggableRows;
            },
            set: function (aValue) {
                if (draggableRows !== aValue) {
                    draggableRows = aValue;
                }
            }
        });
        Object.defineProperty(this, 'keyboardSelectedColumn', {
            get: function () {
                return keyboardSelectedColumn;
            },
            set: function (aValue) {
                if (keyboardSelectedColumn !== aValue) {
                    keyboardSelectedColumn = aValue;
                }
            }
        });
        Object.defineProperty(this, 'keyboardSelectedRow', {
            get: function () {
                return keyboardSelectedRow;
            },
            set: function (aValue) {
                if (keyboardSelectedRow !== aValue) {
                    keyboardSelectedRow = aValue;
                }
            }
        });

        function insertColumn(index, aColumn, needRedraw) {
            if (arguments.length < 3)
                needRedraw = true;
            if (index >= 0 && index <= columns.length) { // It is all about insertBefore
                if (index < columns.length) { // It is all about insertBefore
                    columns.splice(index, 0, aColumn);
                    var col = colgroup.getChild(index);
                    colgroup.insertBefore(aColumn.addCol(), col);
                } else {
                    columns.push(aColumn);
                    colgroup.appendChild(aColumn.addCol());
                }
                table.parentElement.parentElement.appendChild(aColumn.columnRule);
                if (needRedraw) {
                    redraw();
                }
            }
        }
        Object.defineProperty(this, 'insertColumn', {
            get: function () {
                return insertColumn;
            }
        });
        function addColumn(aColumn, needRedraw) {
            if (arguments.length < 2)
                needRedraw = true;
            insertColumn(columns.length, aColumn, needRedraw);
        }
        Object.defineProperty(this, 'addColumn', {
            get: function () {
                return addColumn;
            }
        });

        function removeColumn(index, needRedraw) {
            if (arguments.length < 2)
                needRedraw = true;
            if (index >= 0 && index < columns.length) {
                var removed = columns.splice(index, 1)[0];
                removed.elements.forEach(function (col) {
                    col.parentElement.removeChild(col);
                });
                removed.elements.splice(0, removed.elements.length);
                if (removed.columnRule.parentElement)
                    removed.columnRule.parentElement.removeChild(removed.columnRule);
                if (needRedraw) {
                    redraw();
                }
                return removed;
            } else {
                return null;
            }
        }
        Object.defineProperty(this, 'removeColumn', {
            get: function () {
                return removeColumn;
            }
        });

        Object.defineProperty(this, 'columnsCount', {
            get: function () {
                return columns.length;
            }
        });

        function getColumn(index) {
            return index >= 0 && index < columns.length ? columns[index] : null;
        }
        Object.defineProperty(this, 'getColumn', {
            get: function () {
                return getColumn;
            }
        });

        function getColumnIndex(column) {
            return columns.indexOf(column);
        }

        Object.defineProperty(this, 'rowsCount', {
            get: function () {
                return tbody.rows.length;
            }
        });

        function getViewCell(aRow, aCol) {
            var viewRows = tbody.rows;
            if (aRow - startRow >= 0 && aRow - startRow < viewRows.length) {
                var viewRow = viewRows[aRow];
                var cells = viewRow.cells;
                if (aCol >= 0 && aCol < cells.length) {
                    return cells[aCol];
                }
            }
            return null;
        }
        Object.defineProperty(this, 'getViewCell', {
            get: function () {
                return getViewCell;
            }
        });

        function focusCell(aRow, aCol) {
            var cell = getViewCell(aRow, aCol);
            if (cell) {
                keyboardSelectedColumn = aCol;
                keyboardSelectedRow = aRow;
                cell.focus();
            }
        }
        Object.defineProperty(this, 'focusCell', {
            get: function () {
                return focusCell;
            }
        });

        Object.defineProperty(this, 'keyboardSelectedElement', {
            get: function () {
                return keyboardSelectedElement;
            }
        });

        function redraw() {
            redrawHeaders();
            redrawBody();
            redrawFooters();
        }
        Object.defineProperty(this, 'redraw', {
            get: function () {
                return redraw;
            }
        });

        function recreateHead() {
            if (thead && thead.parentElement)
                table.removeChild(thead);
            if (colgroup && colgroup.parentElement)
                table.removeChild(colgroup);
            thead = document.createElement('thead');
            table.insertBefore(thead, tbody);
            table.insertBefore(colgroup, thead);
        }

        function redrawHeaders() {
            recreateHead();
            drawHeaders();
        }
        Object.defineProperty(this, 'redrawHeaders', {
            get: function () {
                return redrawHeaders;
            }
        });

        function drawHeaders() {
            if (columns.length > 0) {
                var r = 0;
                var nextLayer = headerNodes;
                while (nextLayer.length > 0) {
                    nextLayer = drawHeaderRow(nextLayer);
                    r++;
                }
                while (r < headerMaxDepth) {
                    drawHeaderRow([]);
                    r++;
                }
            }
        }

        function drawHeaderRow(layer) {
            var children = [];
            var tr = document.createElement('tr');
            tr.className = 'p-grid-header-row ' + dynamicHeaderRowsClassName;
            layer.forEach(function (node) {
                tr.appendChild(node.view.element);
                node.view.element.className = 'p-grid-header-cell ' + dynamicHeaderCellsClassName; // reassign classes
                node.view.element.classList.add(node.column.styleName);
                Array.prototype.push.apply(children, node.children);
            });
            thead.appendChild(tr);
            return children;
        }

        function recreateBody() {
            if (tbody && tbody.parentElement)
                table.removeChild(tbody);
            tbody = document.createElement('tbody');
            table.insertBefore(tbody, tfoot);
        }

        function redrawBody() {
            recreateBody();
            drawBody();
        }

        Object.defineProperty(this, 'redrawBody', {
            get: function () {
                return redrawBody;
            }
        });

        function drawBody() {
            if (columns.length > 0) {
                drawBodyPortion(startRow, endRow);
            }
        }

        function drawBodyPortion(start, end) {
            if (end - start > 0) {
                if (!bodyFiller.parentElement)
                    table.parentElement.appendChild(bodyFiller);
                for (var i = start; i < end; i++) {
                    var dataRow = data[i];
                    var viewRow = document.createElement('tr');
                    viewRow.className = 'p-grid-row ' + dynamicRowsClassName;
                    if ((i + 1) % 2 === 0) {
                        viewRow.classList.add(dynamicEvenRowsClassName);
                    } else {
                        viewRow.classList.add(dynamicOddRowsClassName);
                    }
                    if (grid.isSelected(dataRow))
                        viewRow.classList.add('p-grid-selected-row');
                    viewRow[JS_ROW_NAME] = dataRow;
                    if (i < startRow) {
                        // insertFirst ...
                        if (tbody.firstElementChild)
                            tbody.insertBefore(viewRow, tbody.firstElementChild);
                        else
                            tbody.appendChild(viewRow);
                        // ... insertFirst
                    } else {
                        tbody.appendChild(viewRow);
                    }
                    for (var c = 0; c < columns.length; c++) {
                        var column = columns[c];
                        var viewCell = document.createElement('td');
                        // TODO: Check alignment of the cell
                        // TODO: Check image decoration of the cell and decoration styles
                        viewCell.className = 'p-grid-cell ' + dynamicCellsClassName;
                        viewRow.appendChild(viewCell);
                        column.render(i, dataRow, viewCell);
                    }
                }
                var rowSpace = endRow - startRow;
                var renderedRowPortion = end - start;
                var fillerHeight = rowsHeight * (data ? rowSpace - renderedRowPortion : 0);
                bodyFiller.style.height = fillerHeight + 'px';
                bodyFiller.style.display = fillerHeight === 0 ? 'none' : '';
            }
        }

        function setRange(start, end, needRedraw) {
            if (arguments.length < 3)
                needRedraw = true;
            startRow = start;
            endRow = end;
            if (needRedraw) {
                redrawBody();
            }
        }
        Object.defineProperty(this, 'setRange', {
            get: function () {
                return setRange;
            }
        });

        function shiftRange(aStartRow, aEndRow) {
            if (aStartRow >= endRow || aEndRow < startRow) {
                setRange(aStartRow, aEndRow);
            } else {
                if (aStartRow < startRow) {
                    drawBodyPortion(aStartRow, startRow);
                } else if (aStartRow > startRow) {
                    for (var i1 = startRow; i1 < aStartRow; i1++) {
                        tbody.removeChild(tbody.rows[0]);
                    }
                }
                startRow = aStartRow;
                if (aEndRow < endRow) {
                    for (var i2 = endRow - 1; i2 >= aEndRow; i2--) {
                        tbody.removeChild(tbody.rows[tbody.rows.length - 1]);
                    }
                } else if (aEndRow > endRow) {
                    drawBodyPortion(endRow, aEndRow);
                }
                endRow = aEndRow;
            }
        }
        Object.defineProperty(this, 'shiftRange', {
            get: function () {
                return shiftRange;
            }
        });

        function recreateFoot() {
            if (tfoot && tfoot.parentElement)
                table.removeChild(tfoot);
            tfoot = document.createElement('tfoot');
            table.appendChild(tfoot);
        }

        function redrawFooters() {
            recreateFoot();
            drawFooters();
        }
        Object.defineProperty(this, 'redrawFooters', {
            get: function () {
                return redrawFooters;
            }
        });

        function drawFooters() {
            if (columns.length > 0) {
            }
        }

        function redrawRow(index) {
            if (index - startRow >= 0 && index - startRow < tbody.rows.length) {
                var viewRow = tbody.rows[index - startRow];
                var dataRow = data[index];
                for (var c = 0; c < columns.length; c++) {
                    var column = columns[c];
                    var viewCell = viewRow.cells[c];
                    column.render(index, dataRow, viewCell);
                }
            } // if the row is not rendered then we have nothing to do
        }

        function redrawColumn(aIndex) {
            if (aIndex >= 0 && aIndex < columns.length) {
                var column = columns[aIndex];
                for (var i = startRow; i < endRow; i++) {
                    var dataRow = data[i];
                    var viewRow = tbody.rows[i - startRow];
                    var viewCell = viewRow.cells[aIndex];
                    while (viewCell.firstElementChild) {
                        viewCell.removeChild(viewCell.firstElementChild);
                    }
                    column.render(i, dataRow, viewCell);
                }
            }
        }

        function setHeaderNodes(aHeader, maxDepth, needRedraw) {
            if (arguments.length < 2)
                needRedraw = false;
            headerNodes = aHeader;
            headerMaxDepth = maxDepth;
            if (needRedraw) {
                redrawHeaders();
            }
        }
        Object.defineProperty(this, 'setHeaderNodes', {
            get: function () {
                return setHeaderNodes;
            }
        });

        function clearColumnsAndHeader(needRedraw) {
            if (arguments.length < 1)
                needRedraw = true;
            columns.forEach(function (removed) {
                removed.elements.forEach(function (col) {
                    col.parentElement.removeChild(col);
                });
                removed.elements.splice(0, removed.elements.length);
                if (removed.columnRule.parentElement)
                    removed.columnRule.parentElement.removeChild(removed.columnRule);
            });
            columns = [];
            headerNodes = [];
            if (needRedraw) {
                redraw();
            }
        }
        Object.defineProperty(this, 'clearColumnsAndHeader', {
            get: function () {
                return clearColumnsAndHeader;
            }
        });
    }
    return Section;
});